Location selector: Move location on map click

karwosts opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

~~Add an icon-button to ha-selector-location which drops the marker at the center of the current viewport.~~

Move marker on click for location selector.

This is much faster than the old method (which I find very cumbersome), where you have to:
 - zoom out so you can see the current marker location and target location
 - drag the marker to the roughly desired area
 - zoom in 
 - refine marker location
 - repeat zoom/refine

## Type of change
<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

karwosts committed:
Add button to location-selector to move marker to current view

MindFreeze commented:
Thanks for bringing this up. The team discussed this and we decided a better UX would be to move the marker on single press. So you tap/click the map anywhere to place the marker there. Obviously this shouldn't interfere with panning the map or resizing the circle.
Are you interested in implementing this @karwosts ?

karwosts commented:
I poked at this for a bit but so far I've been unsuccessful in figuring out how to generate a click event on the base map. I may come back to it to try again sometime but I'm not sure. 

Right now the map shows the grab cursor on hover, would it be changed to the pointer cursor instead?

MindFreeze commented:
> Right now the map shows the grab cursor on hover, would it be changed to the pointer cursor instead?

@matthiasdebaat what do you think?

matthiasdebaat commented:
> Right now the map shows the grab cursor on hover, would it be changed to the pointer cursor instead?

Yes, I think this is needed to make this interaction work. On drag it should switch to the grab cursor.

We have to check if this also works on mobile. Alternative for touch interfaces could be to have the marker in a fixed center position of the map, allowing users to drag the map underneath instead of dragging the marker itself.

wendevlin commented:
Hi @karwosts,

you can add event listeners to the leaflet map object in `ha-map.ts`: https://leafletjs.com/reference.html#map-click

`this.leafletMap.on("click", () => { fireEvent(this, "click") })` 

then you can handle the click somewhere.
And leaflet will handle that it's not a drag or something else.

karwosts committed:
move on click

karwosts commented:
Ok I've changed the approach as requested.  Removed the new button, and now inside the ha-selector-location, clicking on the map moves the target to the location of the click. 

It also seemed to work as expected on my Android mobile as well (tap moves the marker, hold pans the map).

wendevlin commented:
Great thanks, I can test it tomorrow on my iPhone.

MindFreeze commented:
LGTM. Will leave it for @wendevlin to merge after trying iOS

wendevlin commented on src/components/map/ha-locations-editor.ts:
```diff
@@ -199,15 +204,21 @@ export class HaLocationsEditor extends LitElement {
-    const marker = ev.target;
-    const latlng: LatLng = marker.getLatLng();
-    let longitude: number = latlng.lng;
+  private normalizeLongitude(longitude: number): number {
```
```suggestion
  private _normalizeLongitude(longitude: number): number {
```

MindFreeze replied:
btw. I don't like this style personally but if we are going to enforce it, we should just a lint rule for it

wendevlin commented on src/components/map/ha-locations-editor.ts:
```diff
@@ -199,15 +204,21 @@ export class HaLocationsEditor extends LitElement {
+    const latlng: LatLng = marker.getLatLng();
+    const location: [number, number] = [
+      latlng.lat,
+      this.normalizeLongitude(latlng.lng),
```
```suggestion
      this._normalizeLongitude(latlng.lng),
```

wendevlin commented on src/components/map/ha-locations-editor.ts:
```diff
@@ -64,6 +64,9 @@ export class HaLocationsEditor extends LitElement {
   public themeMode: ThemeMode = "auto";
 
+  @property({ type: Boolean, attribute: "move-on-click" })
+  public moveOnClick = false;
```
this naming could be confusing.
What about

- updateOnClick
- pinOnClick
- ?

karwosts replied:
pinOnClick sounds good, I hesitated with this current name as well.

wendevlin commented on src/components/map/ha-map.ts:
```diff
@@ -191,6 +202,11 @@ export class HaMap extends ReactiveElement {
     try {
       [this.leafletMap, this.Leaflet] = await setupLeafletMap(map);
       this._updateMapStyle();
+      this.leafletMap.on("click", (ev) => {
```
we need to avoid click firing on double click.
Check here: https://stackoverflow.com/questions/29035896/leaflet-dont-fire-a-click-event-function-on-double-click

MindFreeze replied:
Do we? I don't see a problem with double click doing the same as single

wendevlin replied:
Double click is zoom in. So when you zoom in it sets the pin now. It should just set the pin on single click.

wendevlin commented on src/components/map/ha-locations-editor.ts:
```diff
@@ -232,6 +243,22 @@ export class HaLocationsEditor extends LitElement {
+      ];
+      fireEvent(this, "location-updated", { id, location }, { bubbles: false });
+
+      // If the normalized longitude wraps around the globe, pan to the new location.
+      if (location[1] !== ev.detail.location[1]) {
+        this.map.leafletMap?.panTo({ lat: location[0], lng: location[1] });
+      }
```
Can you explain what this block is doing? When does it occure?

wendevlin replied:
Understood thanks

karwosts replied:
The issue was if you zoom all the way out of the map, you see a continuously wrapping globe, and you can zoom back in to anywhere. So if you zoom into say the right side of the map you could be zoomed into lng = 450 - 470. Then when you click a pin at lng=460, it would get normalized to lng=100, and you wouldn't be able to see it. So it would seem to be broken.

wendevlin commented:
> Great thanks, I can test it tomorrow on my iPhone.

iOS works fine :heavy_check_mark:

karwosts committed:
double-click handling

karwosts commented:
Added double-click handling.
